Add differential expression

asizemore opened the pull request:
This PR adds the differential expression frontend plugin.

The differential expression plugin is intentionally copy/pasted from differential abundance. My two reasons for this are:
1. Prevent accidentally breaking mbio during differential expression development
2. The needs for this differential expression app are still evolving, and this path allows us the shortest time to a mvp while giving us space for that growth. If it turns out that in 6 months diff abund and diff expression are the same and we don't see them diverging any more, then we can go merge them together.


Two things to note:
1. Currently we can only test with mbio data. The backend has a temporary hack that transforms the loaded mbio data into what looks like RNA-Seq data. 
2. I left the `collectionPredicate` as `true` to allow all collections to be available in the app. We need to change this once we have real data to work with and we know what types of annotations we'll have.


Testing
- To test the app in the workspace, use /eda or /mbio-site
- To test in the notebook, test with /eda, not /mbio-site or /genomics-site.

dmfalke reviewed (commented):
I think your rationale for copying differential abundance makes sense. The layout of "Group Comparison" looks funny, in your screenshot, but I don't see anything specific to this PR that would cause that. Just wanted to mention it.

I haven't tested anything yet, but hope to soon. I suppose there isn't any harm in merging this, as the option won't show up until it's included in the eda data.

asizemore commented:
Excellent. Thanks @dmfalke ! I'll also look into that "Group Comparison" funky layout.

The backend has been merged, so we can now use the qa mbio site for testing. The backend currently will show this app for genomics (the intended user) and mbio (only used right now for testing because we dont have the genomcis data ready). So my current plan is to not merge this PR until I can update the backend and remove differential expression from mbio. We could probably have both merged but with irregular releases i don't want to miss something rolling out and accidentally taking this change with it!

asizemore commented:
Just updated with `main`. Reminder we're keeping all the differential expression frontend work on a branch until we have data in the correct db. This way we don't accidentally release this app for mbio :)

asizemore commented:
Updated and fixed layout issue.

Note, I directly copy/pasted the differential abundance css and it was quite painful to do so. I went back and forth between merging the two styles and sticking to the principle of separating new work from mbio work. The principle won in my head, but perhaps i've gone too far...

Still works!
<img width="888" alt="Screen Shot 2025-02-21 at 4 58 48 PM" src="https://github.com/user-attachments/assets/599f67b2-6172-4afc-8851-d6596e4c9ffd" />

asizemore commented:
Note to @dmfalke and @jbrestel - we should wait to merge this branch until the data is ready. Currently our only way to test this feature is with mbio data, and I don't want to merge then accidentally release this feature for mbio.

asizemore commented:
Now if the subset is changed, the Group A and B values will reset. There's a funny bug (you can see in the video) when i change the subset then change the group A values too quickly that the new values don't stick. Still a good step forward.

https://github.com/user-attachments/assets/649e7738-5bce-45d6-a447-b97055108553

asizemore commented:
I messed around with where the snackbar should be called. I originally wanted it in the subsetting cell, but then getting all the dependencies correct made the subsetting cell to involved in the compute cell's life, so i didn't like that. The differential abundance config was getting reset in the app plugin anyways, so i moved the enqueueSnackbar there. I'd still like to extract out that logic and make it something we can set in the notebookPreset so that we don't have notebook-specific code in the app. This app won't be in the workspace anytime soon (i think?) so it's not the biggest deal, but even a little isNotebook flag would be okay (i think we did something similar for the map?).

bobular commented on packages/libs/eda/src/lib/notebook/TextNotebookCell.tsx:
```diff
@@ -24,6 +24,7 @@ export function TextNotebookCell(props: NotebookCellProps<TextCellDescriptor>) {
           className={'NotebookCellContent' + (isDisabled ? ' disabled' : '')}
         >
           {text}
+          {getText && getText(analysisState)}
```
`volcanoPlotConfig` has an implicit dependency on `analysisState` through the chain of the two `useMemo`s

But note that most updates to big “state tree” objects are done with the spread operator, e.g.

```ts
setAnalysisState(prev => ({
  ...prev,
  somethingUnrelated: newValue,
}));
```

That pattern creates a **new** `analysisState` object (so its reference always changes), but it **reuses the existing nested objects** unless they’re explicitly replaced. So things like

```ts
prev.analysis?.descriptor.computations[0]
  .visualizations[0].descriptor.configuration
```

keep the **same object identity** if they weren’t touched by the update.

In the “2-memo” version:

```ts
const volcanoPlotConfig = useMemo(() => {
  const config = analysisState.analysis?.descriptor
    .computations[0]?.visualizations[0]?.descriptor?.configuration;
  return VolcanoPlotConfig.is(config) ? config : null;
}, [analysisState]);

const dynamicContent = useMemo(
  () => getText?.(volcanoPlotConfig),
  [getText, volcanoPlotConfig]
);
```

what happens is:

* The first `useMemo` *does* re-run whenever `analysisState` changes, but it’s just doing property access (cheap) and a type guard.
* The second `useMemo` only re-runs when `getText` **or the `volcanoPlotConfig` reference** changes.

Because of the spread/immutability pattern above, `volcanoPlotConfig` is often **the exact same object** across many `analysisState` updates (when the configuration hasn’t changed). In those cases:

* `analysisState` reference changes ✅
* `volcanoPlotConfig` reference stays the same ✅
* `dynamicContent`’s `useMemo` does **not** re-run ✅

Whereas in the original version that depended directly on `analysisState`:

```ts
const dynamicContent = useMemo(
  () => getText?.(analysisState),
  [getText, analysisState]
);
```

any change to anything in `analysisState` forces `dynamicContent` to be recomputed, even if the relevant nested config is unchanged.

So we’re still “technically” dependent on `analysisState` in the selector memo, but we’re **shielding the expensive `getText` work** behind a dependency (`volcanoPlotConfig`) that only changes when the specific nested config actually changes.
